storage: real-time recency MVP

sploiselle opened the pull request:
MVP for real-time recency. Once this merges, it should be appropriate for private preview.

### Motivation

This PR adds a known-desirable feature. #16060

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

sploiselle commented:
@MaterializeInc/testing Could use y'all's expertise in developing a testing suite against this prototype of real-time recency. The idea behind RTR is that when you issue a RTR query, we wait until we've ingested all data that is visible in the external system at the "same" physical time that the query gets issued.

Ideally, we would:
1. Introduce a delay in ingesting it simulating ingestion lag (but not a transient network error from the broker)
2. Produce Kafka data
3. Issue a non-RTR query and show the data is not present
4. Issue a RTR query and show that the data is present.

The delay we introduce would need to be long enough that we can convince ourselves that we're not just "getting lucky."

Which testing framework would you recommend for this? My guess is instrumenting the Kafka source with a `failpoint` to introduce the latency and then everything else could be done in testdrive.

One small wrinkle with that approach is that I would ultimately like to query multiple Kafka sources, each with their own distinct latency and getting that wired up with `failpoint` seems fussy, though maybe it's doable?

def- commented:
For the ingestion lag toxiproxy is good, we already use it in some tests, you can make it as slow as you want in ingesting.
Testdrive seems fine for the queries. I think the kafka-resumption tests in test/kafka-resumption/mzcompose.py come relatively close to this.

> One small wrinkle with that approach is that I would ultimately like to query multiple Kafka sources, each with their own distinct latency and getting that wired up with failpoint seems fussy, though maybe it's doable?

This is doable with Toxiproxy, you can have multiple Kafka sources, all with their own delays, see for example workflow_sink_networking.

If this is too cumbersome, I can also write a test for you.

sploiselle commented:
@def- This is in a shape that we can begin testing it. The RTR operation itself is expensive, so opening 100 (1000? 10?) clients doing simultaneous RTR queries might fall over in some unexpected way. The most meaningful tests will ensure we uphold our semantics, but performance is a secondary concern.

I'm going to work on seeing if there's a way to make this the default for testing queries from Kafka sources but that seems a little tricky.

sploiselle commented:
@nrainer-materialize Here's the real-time recency work we chatted about expanding the tests of. I believe @def- had some outstanding tests, but am not sure exactly what he'd planned or had in mind.

nrainer-materialize commented:
> @nrainer-materialize Here's the real-time recency work we chatted about expanding the tests of. I believe @def- had some outstanding tests, but am not sure exactly what he'd planned or had in mind.

Thanks! I will have a look and work on it next week.

sploiselle commented:
@rjobanp You have bandwidth/interest in adding a MySQL real-time recency test akin to the Kafka and PG ones? I'm sure you're much defter than I with their dialect of SQL, so thought I'd ask. nbd if you're bandwidth constrained.

rjobanp commented:
@sploiselle based on the pg-rtr test it should be almost identical for you to add a mysql one -- the only change I that `generate_series` doesn't exist, but you can use the `@i:=@i+1` pattern we have here: https://github.com/rjobanp/materialize/blob/766595589acdf619512ac8057b5df5cfb7c40187/test/mysql-cdc/mzcompose.py#L164

nrainer-materialize commented:
Oh, I do get some errors for https://github.com/MaterializeInc/materialize/pull/25566 in https://buildkite.com/materialize/nightly/builds/7683. Though, I haven't had yet the time to take a deeper look.

nrainer-materialize commented:
Alright. Regarding testing:

### RTR enabled everywhere
https://github.com/MaterializeInc/materialize/pull/25463, originally by Dennis, enables RTR in all CI. This PR is a one-shot test and **not supposed to be merged.**

#### [Tests pipeline](https://buildkite.com/materialize/test/builds/81756)
* [MySQL CDC](https://buildkite.com/materialize/test/builds/81768#018f5d7f-867e-4390-b8eb-e31301ba8ab1) and [Postgres CDC](https://buildkite.com/materialize/test/builds/81768#018f5d7f-6f1a-4b8f-b326-e37fccb80bbc) fail due to a result mismatch. The test receives more values than expected. Both failures persist after a retry.
* For this test run, [I increased the timeout](https://github.com/MaterializeInc/materialize/pull/25463/commits/f5b8bb2b4604d231cd496b61b48f3d609849a757) of different tests, including testdrive tests. Nonetheless, 5 of the 8 sharded runs time out. (Original timeout: 40 minutes per sharded instance, average run time: about 10-13 minutes per sharded instance, adjusted timeout: 80 minutes per sharded instance) => There is a major performance impact!
* https://github.com/MaterializeInc/materialize/pull/25566/commits/f0e295a25044eb66e10279b19ab07c0ae7ce957b will fix one testdrive failure.

#### [Nightly](https://buildkite.com/materialize/nightly/builds/7720)
* The feature benchmark fails in scenario `StartupLoaded`, which is three times slower. However, I am actually surprised that we don't see more performance regressions here, given the testdrive build timeouts in the tests pipeline.
* Testdrive build time out here as well, although [I had increased timeouts](https://github.com/MaterializeInc/materialize/pull/25463/commits/b3a505af2f88d7b540e71b7e295d58e8d34b279d) for nightly as well.
* Four parallel workload tests fail due to `timed out before ingesting the source\'s visible frontier when real-time-recency query issued`.

### Additional tests
https://github.com/MaterializeInc/materialize/pull/25566, originally by Dennis, adds additional RTR tests and extends existing test frameworks (data-ingest, parallel-workload). That branch is based on the one of this PR. All [additional commits of that branch](https://github.com/MaterializeInc/materialize/pull/25566/commits) should be cherry-picked to this branch (except for https://github.com/MaterializeInc/materialize/pull/25566/commits/719ac3096a3582cbec08caf6f78f9e2f93875ffb, which was cherry-picked from main and could cause conflicts).

#### [Added tests in tests pipeline](https://buildkite.com/materialize/test/builds/81755)
Of the added tests, there are some that do not pass but presumably should. They are in workflow resumption in `kafka-rtr`, manipulate the connection, fail due to `BrokerTransportFailure (Local: Broker transport failure)`, and were disabled with https://github.com/MaterializeInc/materialize/pull/25566/commits/80714258569f1e3c120a3646da62b3b63026ba1f. Consequently, the build step with the remaining workflow passes.

#### Nightly
The nightly pipeline (nearly) passes after adjusting some timeouts.
* Original build: https://buildkite.com/materialize/nightly/builds/7683
* Follow-up 1: https://buildkite.com/materialize/nightly/builds/7702 (the parallel-workload was seen and fixed on main, should be gone after a rebase)
* Follow-up 2: https://buildkite.com/materialize/nightly/builds/7705

sploiselle commented:
@nrainer-materialize Thank you _SO_ much for the detailed reporting here. The nightly timeouts make sense to me but I'm pretty spooked by the result mismatches

nrainer-materialize commented:
@sploiselle, please let me know if you need further support or want me to retest something. Thank you.

sploiselle commented:
@nrainer-materialize Will do! I'm making my way through the feedback you provided and so far all of the issues have been either issues of scale or subtle issues with the tests themselves. Planning a detailed accounting of what's to be done, but I'm feeling good that this prototype is in the right shape.

sploiselle commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -1994,23 +1994,57 @@ impl Coordinator {
+            // data. We "cheat" a little bit and filter out any IDs that aren't
+            // user objects because we know they are not a RTR source.
+            let mut to_visit = VecDeque::from_iter(source_ids.filter(GlobalId::is_user));
+            // If none of the sources are user objects, we don't support RTR.
```
err--this should be more like "we don't need to provide a RTR timestamp"

sploiselle commented on misc/python/materialize/parallel_workload/action.py:
```diff
@@ -232,6 +232,7 @@ def errors_to_ignore(self, exe: Executor) -> list[str]:
             [
                 "in the same timedomain",
                 'is not allowed from the "mz_introspection" cluster',
+                "timed out before ingesting the source's visible frontier when real-time-recency query issued",
```
This got inadvertently brought over from testing; will remove.

sploiselle commented on misc/python/materialize/parallel_workload/action.py:
```diff
@@ -288,7 +289,11 @@ def run(self, exe: Executor) -> bool:
 
         query += " LIMIT 1"
 
+        if isinstance(obj, KafkaSource) and (not join or isinstance(obj2, KafkaSource)):
```
This got inadvertently brought over from testing; will remove.

sploiselle commented on misc/python/materialize/data_ingest/workload.py:
```diff
@@ -23,7 +23,12 @@
     RecordSize,
     Upsert,
 )
-from materialize.data_ingest.executor import PgExecutor, PrintExecutor
```
All diffs in this file inadvertently brought over from testing; will remove.

sploiselle commented on src/adapter/src/coord/message_handler.rs:
```diff
@@ -841,6 +842,39 @@ impl Coordinator {
+                        let name = conn_catalog
+                            .minimal_qualification(conn_catalog.get_item(&id).name())
+                            .to_string();
+                        crate::AdapterError::RtrTimeout(name)
```
This should be `RtrDropFailure`

maddyblue replied:
This is still RtrTimeout. Is that correct?

sploiselle commented on src/storage-controller/Cargo.toml:
```diff
@@ -30,14 +30,23 @@ mz-repr = { path = "../repr" }
 serde_json = { version = "1.0.89" }
-timely = { version = "0.12.0", default-features = false, features = ["bincode"] }
-tokio = { version = "1.24.2", features = ["fs", "rt", "sync", "test-util", "time"] }
+timely = { version = "0.12.0", default-features = false, features = [
```
No idea why my formatter for `.toml` files is different from everyone else's will spend a few minutes trying to solve that before mergin this.

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -2694,6 +2694,103 @@ where
+                        (connection.clone(), *remap_collection_id)
+                    }
+
+                    // These internal sources will never support RTR.
```
Never is too strong a word; maybe they should support RTR?

bosconi replied:
And lo, it came to pass...

sploiselle reviewed (commented):
A bit of mess to clean up from integrating QA's tests, as well as notes to tighten up.

sploiselle commented:
> The feature comes with a major performance impact.

The performance impact mentioned here is largely (entirely?) related to latency. We expect the latency of RTR queries to exceed those of non-RTR queries commensurate with the speed at which we ingest data and propagate the ingestion through the entire system.

The prototype itself doesn't introduce much undue latency to the queries (i.e. envd cannot figure out much more quickly than listening to the remap shard when a value's been ingested).

> When run in concurrency (parallel-workload), the problem timed out before ingesting the source\'s visible frontier when real-time-recency query issued may occur.

The scalability of the prototype is not great! We open a separate connection to the upstream object for each query. I expect the number of concurrent queries we support to be quite low, and this is something that we can both get customer feedback on, as well as easily improve. (e.g. an easy win is to only allow one outstanding RTR query per connection object, and we can stash all queries waiting on that object in a queue. Once the current query returns, we can issue _another_ RTR timestamp fetch, and the timestamp returned is valid for all queued queries).

> Resumption tests with this feature were not successful.

The resumption tests are, unfortunately, pathologically structured and can never succeed. One of the tenants of the current prototype is that we use the same connection object as the source itself (i.e. it must have the same parameters, etc.). The current design of the resumption tests introduces a partition (of sorts) between envd and the source's upstream object, and then issues a RTR query. Unfortunately, this partition not only affects the ingestion, but also impedes our ability to determine the RTR timestamp (which is why all of the queries fail).

A trickier design here would be to issues the RTR query and _then_ introduce the issue and _then_ fix it. The timing of all of this seems very racy, though so I'm not sure exactly how we'd want to instrument it.

sploiselle commented:
@MaterializeInc/storage This is ready for a code review. Note that this doesn't include a `mysql-rtr` test, though it probably should. QA tested that the existing tests do pass with RTR enabled and this is only moving to PrPr, so I didn't sweat it too profusely.

nrainer-materialize commented:
> @MaterializeInc/storage This is ready for a code review. Note that this doesn't include a `mysql-rtr` test, though it probably should. QA tested that the existing tests do pass with RTR enabled and this is only moving to PrPr, so I didn't sweat it too profusely.

I can add some if we consider that important!

nrainer-materialize commented:
> > @MaterializeInc/storage This is ready for a code review. Note that this doesn't include a `mysql-rtr` test, though it probably should. QA tested that the existing tests do pass with RTR enabled and this is only moving to PrPr, so I didn't sweat it too profusely.

@sploiselle: I added mysql-rtr tests with https://github.com/MaterializeInc/materialize/pull/25195/commits/06b63ed44c1b6412b3a7a0362d8d76f42c8e7d05.

bosconi commented on test/pg-rtr/rtr/verify-rtr.td:
```diff
@@ -0,0 +1,69 @@
+INSERT INTO table_a SELECT 1,2 FROM generate_series(1, 100);
+INSERT INTO table_b SELECT 1,2 FROM generate_series(1, 100);
+
+> SELECT sum < 4000207 FROM sum;
```
```suggestion
> SELECT sum < 1007 FROM sum;
```

sploiselle replied:
Ty!

bosconi commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -2016,22 +2016,27 @@ impl Coordinator {
             // user objects because we know they are not a RTR source.
             let mut to_visit = VecDeque::from_iter(source_ids.filter(GlobalId::is_user));
-            let mut visited = BTreeSet::new();
+            // If none of the sources are user objects, we don't support RTR.
```
```suggestion
            // If none of the sources are user objects, we don't need to provide a RTR timestamp.
```
What happens in the caller though? The query returns immediately? Or times out because it can't get a timestamp to wait for?

bosconi replied:
Do we have a test case for this? If not, could one be added easily?

sploiselle replied:
> What happens in the caller though? The query returns immediately? Or times out because it can't get a timestamp to wait for?

Their query is not influenced by any external system's frontier, so the RTR machinery just isn't involved in determining a timestamp. We can test this by querying something that doesn't depend on a user source and show that it returns values––certainly doesn't hurt to demonstrate.

bosconi commented on test/kafka-rtr/mzcompose.py:
```diff
@@ -59,8 +59,10 @@ def workflow_simple(c: Composition) -> None:
 
-# TODO: All failure modes fail with: timed out before ingesting the source's visible frontier when real-time-recency query issued. input_1 failed to ingest data up to the real-time recency point
-def workflow_resumption(c: Composition) -> None:
+# It is impossible for this test to succeed because the network failures
```
More accurate to say that this can't pass without making the RTR-related connections robust to network failure?
Is this something that would only make sense to re-visit in the context of the work to add HA/"always on" behavior? Or resuming these broken connections could add value even before HA work is done?
If this is correct, should we open a issue to work on this at some point, and add it to the mega tracker?

sploiselle replied:
For this to work, we would have to do something other than blithely partition `envd` away from the external system. If you partition `envd` away from the external system, you cannot reach out to it to determine its frontier. For this test to work, we would need to do something trickier, like only partition us away from the external system _after_ the thread performing the RTR determination received a response from the query that determines the external frontier.

sploiselle replied:
@MaterializeInc/testing I'm going to remove these functions because as-is they cannot succeed but will leave a TODO for @def- to investigate a way to make them work with a link to the commits in https://github.com/MaterializeInc/materialize/pull/25566 that have the original tests.

nrainer-materialize replied:
Fine with me.

petrosagg commented on src/storage-types/src/sources.rs:
```diff
@@ -838,6 +842,130 @@ pub enum GenericSourceConnection<C: ConnectionAccess = InlinedConnection> {
+                    }
+                    ListenEvent::Progress(f) => {
+                        let updates = Vec::with_capacity(remap_updates.len());
+                        let (max_ts, updates) = remap_updates.drain(..).fold(
```
There are a number of issues with this code. Here we are processing all updates in `remap_updates` even if they are beyond the frontier `f`. Also, we are only considering the maximum timestamp observed in the batch which might produce a non-tight timestamp. Here is a diff with a way to do it that guards against these edge cases:

```diff
diff --git a/src/storage-types/src/sources.rs b/src/storage-types/src/sources.rs
index 222091e7a1..beccac61bf 100644
--- a/src/storage-types/src/sources.rs
+++ b/src/storage-types/src/sources.rs
@@ -9,7 +9,9 @@
 
 //! Types and traits related to the introduction of changing collections into `dataflow`.
 
-use std::collections::BTreeMap;
+use std::cmp::Reverse;
+use std::collections::binary_heap::PeekMut;
+use std::collections::{BTreeMap, BinaryHeap};
 use std::fmt::Debug;
 use std::hash::Hash;
 use std::ops::{Add, AddAssign, Deref, DerefMut};
@@ -44,7 +46,7 @@ use prost::Message;
 use serde::{Deserialize, Serialize};
 use timely::order::{PartialOrder, TotalOrder};
 use timely::progress::timestamp::Refines;
-use timely::progress::{PathSummary, Timestamp};
+use timely::progress::{Antichain, PathSummary, Timestamp};
 
 use crate::configuration::StorageConfiguration;
 use crate::connections::inline::{
@@ -847,6 +849,7 @@ impl GenericSourceConnection {
         self,
         id: GlobalId,
         config: StorageConfiguration,
+        as_of: Antichain<T>,
         remap_subscribe: Subscribe<SourceData, (), T, mz_repr::Diff>,
     ) -> Result<T, StorageError<T>> {
         match self {
@@ -859,6 +862,7 @@ impl GenericSourceConnection {
                 GenericSourceConnection::decode_remap_data_until_geq_external_frontier(
                     id,
                     external_frontier,
+                    as_of,
                     remap_subscribe,
                 )
                 .await
@@ -872,6 +876,7 @@ impl GenericSourceConnection {
                 GenericSourceConnection::decode_remap_data_until_geq_external_frontier(
                     id,
                     external_frontier,
+                    as_of,
                     remap_subscribe,
                 )
                 .await
@@ -885,6 +890,7 @@ impl GenericSourceConnection {
                 GenericSourceConnection::decode_remap_data_until_geq_external_frontier(
                     id,
                     external_frontier,
+                    as_of,
                     remap_subscribe,
                 )
                 .await
@@ -897,11 +903,12 @@ impl GenericSourceConnection {
     }
 
     async fn decode_remap_data_until_geq_external_frontier<
-        ExternalFrontier: SourceTimestamp,
-        T: Timestamp + Lattice + Codec64 + From<EpochMillis>,
+        FromTime: SourceTimestamp,
+        T: Timestamp + Lattice + Codec64,
     >(
         id: GlobalId,
-        external_frontier: timely::progress::Antichain<ExternalFrontier>,
+        external_frontier: timely::progress::Antichain<FromTime>,
+        as_of: Antichain<T>,
         mut remap_subscribe: Subscribe<SourceData, (), T, mz_repr::Diff>,
     ) -> Result<T, StorageError<T>> {
         use mz_persist_client::read::ListenEvent;
@@ -914,55 +921,76 @@ impl GenericSourceConnection {
             external_frontier.pretty()
         );
 
+        let external_frontier = external_frontier.borrow();
         let mut native_upper = MutableAntichain::new();
-        let mut remap_updates = vec![];
 
-        loop {
+        let mut remap_frontier = Antichain::from_elem(T::minimum());
+        let mut pending_remap = BinaryHeap::new();
+
+        let mut min_ts = T::minimum();
+        min_ts.advance_by(as_of.borrow());
+        let mut min_ts = Some(min_ts);
+
+        // First we must read the snapshot so that we can accumulate the minimum timestamp
+        // correctly.
+        while !remap_frontier.is_empty() {
+            // Receive binding updates
             for event in remap_subscribe.fetch_next().await {
                 match event {
                     ListenEvent::Updates(updates) => {
-                        for ((k, v), t, d) in updates {
+                        for ((k, v), into_ts, diff) in updates {
                             let row: Row = k.expect("invalid binding").0.expect("invalid binding");
                             let _v: () = v.expect("invalid binding");
-                            let from_ts: ExternalFrontier = SourceTimestamp::decode_row(&row);
-                            remap_updates.push((from_ts, t, d));
+                            let from_ts: FromTime = SourceTimestamp::decode_row(&row);
+                            pending_remap.push(Reverse((into_ts, from_ts, diff)));
                         }
                     }
-                    ListenEvent::Progress(f) => {
-                        let updates = Vec::with_capacity(remap_updates.len());
-                        let (max_ts, updates) = remap_updates.drain(..).fold(
-                            (T::minimum(), updates),
-                            |(ts, mut updates), (from_ts, t, d)| {
-                                updates.push((from_ts, d));
-                                (std::cmp::max(ts, t), updates)
-                            },
-                        );
-
-                        // If there were no updates, then the ts we get is not
-                        // derived from any remap data.
-                        if updates.is_empty() && !f.is_empty() {
-                            continue;
+                    ListenEvent::Progress(frontier) => remap_frontier = frontier,
+                }
+            }
+            // Only process the data if we can correctly accumulate timestamps beyond the as_of
+            if as_of.iter().any(|t| !remap_frontier.less_equal(t)) {
+                loop {
+                    // The next timestamp that must be considered is either the minimum timestamp,
+                    // if we haven't yet checked it, or the next timestamp of the pending remap
+                    // updates that is not beyond the frontier.
+                    let binding_ts = match min_ts.take() {
+                        Some(min_ts) => min_ts,
+                        None => {
+                            let Some(Reverse((into_ts, _, _))) = pending_remap.peek() else {
+                                break;
+                            };
+                            if !remap_frontier.less_equal(into_ts) {
+                                into_ts.clone()
+                            } else {
+                                break;
+                            }
                         }
-
-                        native_upper.update_iter(updates);
-
-                        // At this TS, the source contained all of the data visible
-                        // in the external system.
-                        if PartialOrder::less_equal(
-                            &external_frontier.borrow(),
-                            &native_upper.frontier(),
-                        ) {
-                            tracing::trace!("real-time recency ts for {id}: {max_ts:?}");
-                            return Ok(max_ts);
-                        } else if f.is_empty() {
-                            // Collection dropped before we ingested external
-                            // frontier.
-                            return Err(StorageError::RtrDropFailure(id));
+                    };
+                    // Create an iterator with all the updates that are less than or equal to
+                    // binding_ts
+                    let binding_updates = std::iter::from_fn(|| {
+                        let update = pending_remap.peek_mut()?;
+                        if PartialOrder::less_equal(&update.0 .0, &binding_ts) {
+                            let Reverse((_, from_ts, diff)) = PeekMut::pop(update);
+                            Some((from_ts, diff))
+                        } else {
+                            None
                         }
+                    });
+                    native_upper.update_iter(binding_updates);
+
+                    // Check if at binding_ts the source contained all of the data visible in the
+                    // external system. If so, we're done.
+                    if PartialOrder::less_equal(&external_frontier, &native_upper.frontier()) {
+                        tracing::trace!("real-time recency ts for {id}: {binding_ts:?}");
+                        return Ok(binding_ts);
                     }
                 }
             }
         }
+        // Collection dropped before we ingested external frontier.
+        Err(StorageError::RtrDropFailure(id))
     }
 }
 
```

sploiselle replied:
Thanks for this; couldn't have come up with this implementation myself.

sploiselle commented on test/pg-rtr/rtr/verify-rtr.td:
```diff
@@ -0,0 +1,69 @@
+INSERT INTO table_a SELECT 1,2 FROM generate_series(1, 100);
+INSERT INTO table_b SELECT 1,2 FROM generate_series(1, 100);
+
+> SELECT sum < 4000207 FROM sum;
```
This file already has that declaration: https://github.com/MaterializeInc/materialize/pull/25195/files/06b63ed44c1b6412b3a7a0362d8d76f42c8e7d05#diff-1fa6e17f2bdbae3013d27b64e932e90f9844adc5ccd5c782038e7d9745c92e9fR24

sploiselle commented on test/mysql-rtr/rtr/verify-rtr.td:
```diff
@@ -0,0 +1,72 @@
+INSERT INTO table_a SELECT 1,2 FROM mysql.time_zone t1, mysql.time_zone t2 LIMIT 100;
+INSERT INTO table_b SELECT 1,2 FROM mysql.time_zone t1, mysql.time_zone t2 LIMIT 100;
+
+> SELECT sum < 4000207 FROM sum;
```
This file already has that declaration: https://github.com/MaterializeInc/materialize/pull/25195/files/06b63ed44c1b6412b3a7a0362d8d76f42c8e7d05#diff-439842c74a0ad8a94839325bf735e01aeb33c82f0636c4b0a47be988017b8967R27

sploiselle commented on test/kafka-rtr/mzcompose.py:
```diff
@@ -0,0 +1,202 @@
+        t.join()
+
+
+def workflow_multithreaded(c: Composition) -> None:
```
@MaterializeInc/testing PTAL

sploiselle commented on test/kafka-rtr/multithreaded/verify-rtr.td:
Looks like this got refactored away by QA at some point; will remove

sploiselle commented on test/kafka-rtr/multithreaded/mz-setup.td:
Looks like this got refactored away by QA at some point; will remove

sploiselle commented on test/kafka-rtr/multithreaded/ingest.td:
Looks like this got refactored away by QA at some point; will remove

sploiselle commented on test/kafka-rtr/simple/ingest.td:
Looks like this got refactored away by QA at some point; will remove

maddyblue commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -1994,23 +1994,57 @@ impl Coordinator {
+
+            let r = self
+                .controller
+                .recent_timestamp(timestamp_objects, *vars.statement_timeout())
```
`statement_timeout` may be wrong here? Its docs say

> Sets the maximum allowed duration of INSERT...SELECT, UPDATE, and DELETE operations.

I'm not opposed to leaving it as is because we just need some timeout, and this isn't a bad one and there's not an obvious alternative.

sploiselle replied:
Yeah, I think you're right and this should just have its own timeout.

maddyblue reviewed (commented):
Adapter parts lgtm.

sploiselle commented on src/storage-types/src/sources.rs:
```diff
@@ -838,6 +842,130 @@ pub enum GenericSourceConnection<C: ConnectionAccess = InlinedConnection> {
+                )
+                .await
+            }
+            s @ GenericSourceConnection::LoadGenerator(_) => unreachable!(
```
I described this in a new `# Panics` section; getting to this point in the code represents a programming error, so think it should be `unreachable`

sploiselle commented on src/storage-types/src/sources/kafka.rs:
```diff
@@ -120,6 +125,60 @@ impl<C: ConnectionAccess> KafkaSourceConnection<C> {
+
+        let meta = client
+            .inner()
+            .fetch_metadata(Some(&self.topic), metadata_timeout)?;
```
Thank you--done.

sploiselle commented on src/storage-types/src/sources/kafka.rs:
```diff
@@ -37,6 +38,10 @@ include!(concat!(
 
+/// The frontier that Kafka presents––for each partition, the greatest visible
+/// offset.
+pub type NativeFrontier = Partitioned<RangeBound<i32>, MzOffset>;
```
Fixed. Thank you.

sploiselle commented on src/storage-types/src/sources.rs:
```diff
@@ -838,6 +842,130 @@ pub enum GenericSourceConnection<C: ConnectionAccess = InlinedConnection> {
+    }
+
+    async fn decode_remap_data_until_geq_external_frontier<
+        ExternalFrontier: SourceTimestamp,
```
Fixed. Thank you.

sploiselle commented on src/storage-types/src/sources.rs:
```diff
@@ -838,6 +842,130 @@ pub enum GenericSourceConnection<C: ConnectionAccess = InlinedConnection> {
+
+    async fn decode_remap_data_until_geq_external_frontier<
+        ExternalFrontier: SourceTimestamp,
+        T: Timestamp + Lattice + Codec64 + From<EpochMillis>,
```
Do you have pointers for concretizing this to `mz_repr::Timestamp`? That the storage controller gets made available through a trait makes this hard to do.

sploiselle commented on src/controller/src/lib.rs:
```diff
@@ -521,14 +521,12 @@ where
-    ) -> BoxFuture<'static, T> {
-        // Dummy implementation
-        Box::pin(async { T::minimum() })
+        ids: BTreeSet<GlobalId>,
```
Clarified the comment

sploiselle commented on src/storage-types/src/sources.rs:
```diff
@@ -838,6 +842,130 @@ pub enum GenericSourceConnection<C: ConnectionAccess = InlinedConnection> {
+    ) -> Result<T, StorageError<T>> {
+        use mz_persist_client::read::ListenEvent;
+        use mz_timely_util::antichain::AntichainExt;
+        use timely::progress::frontier::MutableAntichain;
```
Oh sure––I think that I was having some obnoxious IDE issues when I was doing this originally.

nrainer-materialize commented on src/storage-types/src/sources/mysql.rs:
```diff
@@ -135,6 +135,36 @@ pub static MYSQL_PROGRESS_DESC: Lazy<RelationDesc> = Lazy::new(|| {
         .with_column("transaction_id", ScalarType::UInt64.nullable(true))
 });
 
+impl MySqlSourceConnection {
```
Class not covered by any test according to the [coverage report](https://buildkite.com/materialize/coverage/builds/413)?
Or should this be covered by the MySQL tests that I added?

sploiselle replied:
I inadvertently lost the diff that contained your tests; recovered it and they're back in this PR.

sploiselle commented on src/storage-controller/src/rtr.rs:
```diff
@@ -0,0 +1,194 @@
+    as_of: Antichain<T>,
+    remap_subscribe: Subscribe<SourceData, (), T, mz_repr::Diff>,
+) -> Result<T, StorageError<T>> {
+    match connection {
+        GenericSourceConnection::Kafka(kafka) => {
+            let external_frontier = kafka
+                .fetch_write_frontier(&config)
+                .await
+                .map_err(StorageError::Generic)?;
+
+            decode_remap_data_until_geq_external_frontier(
+                id,
+                external_frontier,
+                as_of,
+                remap_subscribe,
+            )
+            .await
+        }
```
Is it possible the coverage report is mistaken? The Kafka RTR tests should be using this.

nrainer-materialize replied:
Yes, that's definitively an option! The report is not 100% accurate.

sploiselle commented:
@petrosagg Addressed all of your feedback modulo erroring if you have RTR enabled and query a load generator source. I think the ergonomics of that are less friendly than they could be (we let you query everything else that doesn't actually support RTR; making this an exception feels odd to me). I'd prefer to educate users about those semantics in documentation.

Also gave up on semantic commits and everything is just in one commit now. Apologies if that introduces any overhead for you.

sploiselle commented on src/storage-types/src/sources/kafka.rs:
```diff
@@ -120,6 +129,61 @@ impl<C: ConnectionAccess> KafkaSourceConnection<C> {
     }
 }
 
+impl KafkaSourceConnection {
```
Can you double-check my thinking here that these should be covered by the Kafka RTR tests? Glad to hop on a Zoom to hammer this out.

sploiselle commented on src/storage-controller/src/rtr.rs:
```diff
@@ -0,0 +1,194 @@
+            )
+            .await
+        }
+        GenericSourceConnection::MySql(my_sql) => {
+            let external_frontier = my_sql
+                .fetch_write_frontier(&config)
+                .await
+                .map_err(StorageError::Generic)?;
+
+            decode_remap_data_until_geq_external_frontier(
+                id,
+                external_frontier,
+                as_of,
+                remap_subscribe,
+            )
+            .await
```
Ditto https://github.com/MaterializeInc/materialize/pull/25195#discussion_r1607926825

petrosagg reviewed (approved):
Looks good, thanks for addressing the comments
